R/harmonic.R

```diff
@@ -0,0 +1,374 @@
+      rlang::abort("starting_val must be numeric, Date or POSIXt")
+
+
+    f_units <-
+      c("year",
+        "month_synodic",
+        "month_sidereal",
+        "month_average",
+        "week",
+        "day",
+        "hour",
+        "minute",
+        "second",
+        "sample")
+      if (!(cycle_unit %in% f_units)) {
+        rlang::abort(paste0("Possible values of `frequency units` should include: ",
+                     paste0("'", f_units, "'", collapse = ", ")))
+      }
+
+    # allow empty terms
+    terms <- quos(...)
+    if (is_empty(terms)) {
+      terms = NULL
+    }
+
+    add_step(
+      recipe,
+      step_harmonic_new(
+        terms = terms,
+        trained = trained,
+        role = role,
+        frequency = frequency,
+        period = period,
+        cycle_unit = cycle_unit,
+        starting_val = starting_val,
+        objects = objects,
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+step_harmonic_new <-
+  function(terms, role, trained,
+           frequency, period, cycle_unit,
+           starting_val, objects, skip, id) {
+    step(
+      subclass = "harmonic",
+      terms = terms,
+      role = role,
+      trained = trained,
+      frequency = frequency,
+      period = period,
+      cycle_unit = cycle_unit,
+      starting_val = starting_val,
+      objects = objects,
+      skip = skip,
+      id = id
+    )
+  }
+
+
+get_cycle_n <- function(x) {
+  if (x == 'year') 31556926 # solar year
+  else if (x == 'month_synodic') cycle_size = 2551443
+  else if (x == 'month_sidereal') cycle_size = 2360592
+  else if (x == 'month_average') cycle_size = 2629744
+  else if (x == 'week') cycle_size = 86400 * 7
+  else if (x == 'day') cycle_size = 86400
+  else if (x == 'hour') cycle_size = 3600
+  else if (x == 'minute') cycle_size = 60
+  else if (x == 'second') cycle_size = 1
+  else if (x == 'sample') cycle_size = 1
+}
+
+
+process_inputs <- function(time_var,
+                           period = NA,
+                           frequency = NA,
+                           starting_val = NA,
+                           cycle_unit = "week") {
+
+  if (all(is.na(time_var)))
+    rlang::abort("variable must have at least one non-NA value")
+
+
+  # get cycle information
+  cycle_size <- get_cycle_n(cycle_unit)
+  if (inherits(time_var, 'Date')) {
+    cycle_size <- cycle_size / 86400
+  }
+
+  # convert from period to frequency
+  period_1    <- sort(unique(na.omit(period)))
+  frequency_1 <- sort(unique(na.omit(frequency)))
+  frequency_2 <- 1 / period_1
+  frequencies <- c(frequency_1, frequency_2)
+
+  # set column names
+
+  if (length(frequency_1) > 0 & length(period_1) > 0) {
+    nms <- c(paste0('_f_', frequency_1), paste0('_p_', period_1))
+  } else if (length(frequency_1) > 0) {
+    nms <- paste0('_f_', frequency_1)
+  } else if (length(period_1) > 0) {
+    nms <- paste0('_p_', period_1)
+  }
+
+  col_names <- c(paste0(rep("sin", each = length(frequencies)), nms),
```

Comment:
I don't like putting numeric data in column names (esp if it ends up being non-while numbers). Can you propose something different? Even an integer index would be good (with the tidy method giving use the key).

Reply:
Either way, I would have the source data column name first in the new name.

Reply:
I went with your integer suggestion. e.g. varname_sin_1, varname_sin_2, varname_cos_1, varname_cos_2,